samples/subsys/ipc/openamp/src/main.c

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright (c) 2018, NXP
+ * Copyright (c) 2018, Nordic Semiconductor ASA
+ * Copyright (c) 2018, Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <ipm.h>
+#include <misc/printk.h>
+#include <device.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <init.h>
+
+#include <openamp/open_amp.h>
+#include <metal/device.h>
+
+#include "common.h"
+
+#define APP_TASK_STACK_SIZE (1024)
+K_THREAD_STACK_DEFINE(thread_stack, APP_TASK_STACK_SIZE);
+static struct k_thread thread_data;
+
+static struct device *ipm_handle;
+
+static metal_phys_addr_t shm_physmap[] = { SHM_START_ADDR };
+static struct metal_device shm_device = {
+	.name = SHM_DEVICE_NAME,
+	.bus = NULL,
+	.num_regions = 1,
+	{
+		{
+			.virt       = (void *) SHM_START_ADDR,
+			.physmap    = shm_physmap,
+			.size       = SHM_SIZE,
+			.page_shift = 0xffffffff,
+			.page_mask  = 0xffffffff,
+			.mem_flags  = 0,
+			.ops        = { NULL },
+		},
+	},
+	.node = { NULL },
+	.irq_num = 0,
+	.irq_info = NULL
+};
+
+static volatile unsigned int received_data;
+
+static struct virtio_vring_info rvrings[2] = {
+	[0] = {
+		.info.align = VRING_ALIGNMENT,
+	},
+	[1] = {
+		.info.align = VRING_ALIGNMENT,
+	},
+};
+static struct virtio_device vdev;
+static struct rpmsg_virtio_device rvdev;
+static struct metal_io_region *io;
+static struct virtqueue *vq[2];
+
+static unsigned char virtio_get_status(struct virtio_device *vdev)
+{
+	return VIRTIO_CONFIG_STATUS_DRIVER_OK;
+}
+
+static void virtio_set_status(struct virtio_device *vdev, unsigned char status)
+{
+	unsigned char *status_byte = (void *)RSC_TABLE_ADDRESS;
+
+	*status_byte = status;
```

Comment:
how about using sys_write32() as it is to directly access the shared memory between the two processors, and they may not be cache coherent. On the slave side, better to read with sys_read32()?

Reply:
I changed it to sys_{read,write}8